### PR TITLE
[DONT MERGE] 'window_text' update (initial state)

### PR DESCRIPTION
**Pull request by MRprusa3d**
!!! ToDo:
- 'window_text'-parameters extension (incl. 'gui_defaults')
- proper parameters setting for "str2multiline()"

### CODE REVIEW

File: src/guiapi/include/display_helper.h [hunk 1]
```diff
@@ -13,7 +13,7 @@
 extern "C" {
 #endif //__cplusplus
 
-extern void render_text_align(rect_ui16_t rc, const char *text, font_t *font, color_t clr0, color_t clr1, padding_ui8_t padding, uint16_t flags);
+extern void render_text_align(void *context, rect_ui16_t rc, const char *text, font_t *font, color_t clr0, color_t clr1, padding_ui8_t padding, uint8_t align, uint8_t ml_mode);
```
Comment (RadekVana) [inline src/guiapi/include/display_helper.h:16]: consider 2 functions
1 with "extended" signature
1 with original signature (calling extended version)

File: src/guiapi/src/display_helper.c [hunk 2]
```diff
@@ -34,9 +36,15 @@ void render_text_align(rect_ui16_t rc, const char *text, font_t *font, color_t c
         display->fill_rect(rc_l, clr0);
         display->fill_rect(rc_r, clr0);
     } else {
-        point_ui16_t wh_txt = font_meas_text(font, text);
+        char *str_tmp = (char *)0x10000000; // ~ PNG-buffer (CCM RAM)
+        strcpy(str_tmp, text);
+        if (ml_mode == ML_MODE_EXT) {
+            set_instance(((window_text_t *)context)->pml_mode);
+            str2multiline(str_tmp, ((window_text_t *)context)->line_width);
```
Comment (RadekVana) [inline src/guiapi/src/display_helper.c:43]: you should be able to calculate `line_width ` from `font` and `rc`
try pass just `pml_mode` argument instead `context`

File: src/guiapi/include/display_helper.h (hunk 1 above)
Comment (RadekVana) [reply src/guiapi/include/display_helper.h:16]: use `ml_mode_t ml_mode` instead `uint8_t ml_mode`

File: src/common/str_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    count = strlen(pstr);
+    if (n > count)
+        n = count;
+    count = count - n + 1;
+    for (i = 0; i < count; i++, pstr++)
+        *pstr = *(pstr + n);
+    return (n);
+}
+
+size_t strins(char *pstr, const char *pinstr, size_t repeater, bool before_flag) {
```
Comment (RadekVana) [inline src/common/str_utils.cpp:27]: typo strins strings

File: src/common/str_utils.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
+#ifndef _STR_UTILS_H
```
Comment (RadekVana) [inline src/common/str_utils.h:1]: consider `#pragma once` instead `#ifndef` - `#define` - `#endif`

File: src/guiapi/include/guitypes.h [hunk 5]
```diff
@@ -26,6 +26,13 @@
 #define ALIGN_CENTER_TOP    (ALIGN_HCENTER | ALIGN_TOP)
 #define ALIGN_CENTER_BOTTOM (ALIGN_HCENTER | ALIGN_BOTTOM)
 
+//multi-line modes
+typedef enum {
```
Comment (RadekVana) [inline src/guiapi/include/guitypes.h:30]: use 
```
#pragma pack(push)
#pragma pack(1)
```
around `ml_mode_t`
since it is used inside packed structures

**Comment by DRracer** [comment]
closing, not applicable anymore

**Pull request closed without merging**